add test targets in makefiles to run local tests

michelleN opened the pull request.

Commit:
add test targets in makefiles to run local tests

jackgr commented:
go test -v ./... can be run from the root, or from any of the folders in the tree. Why add it to make?

sgoings commented:
This test target looks like it already exists: https://github.com/kubernetes/deployment-manager/blob/master/Makefile#L34

jackgr commented:
Yes, but only at the top level. Same is true for build. The go tools knows how to recurse, and there's nothing to maintain to make it to the right thing. That said, if we have some other kinds of tests to run, then it would make sense to add a test target to the Makefile for the project(s) containing those tests, and to make the test target recursively descend.

sgoings commented on Makefile:
```diff
@@ -48,3 +48,6 @@ $(TARGETS) : % : $(addsuffix %,$(SUBDIRS))
 	$(MAKE) -C $(@D) $(@F:.%=%)
+
+test:
+	go test -v ./...
```
@michelleN this is going to cause an issue due to the fact that it's already defined above.

Commit:
remove duplicate test target @ root

add structure for [lint,vet,test-unit] all wrapped up
by the test target

sgoings commented:
@jackgr if you're following along at home, I just pushed that commit after some time talking to @michelleN about how we could improve the process of [linting, vetting, unit testing, integration testing, ...] in each subproject. @michelleN will be working on it still from here.

Commit:
add golint and go vet to test target

This commit adds a target called setup-gotools which will download and
install golint and go vet. A developer who wants to run tests for an
individual component (manager, resourcifier, expandybird) can run `make
test` to run golint, vet, and test and in that directory. On first run,
a developer will want to run `make setup-gotools` so that they have
golint and vet to use locally.

There is a `make test` command in each component because running tests
from the root directory will result in lots of output. Running tests in
individual components will help isolate the tests for the component that
is being modified and result in more clear output.

michelleN commented:
This PR should satisfy issue #211. @technosophos: Is this what you were thinking?

Commit:
modify travis so it will install golint and govet

* before running tests

sgoings commented on .travis.yml:
```diff
@@ -9,4 +9,5 @@ install:
 
 script:
+  - make setup-gotools
   - make test
```
You could combine these two lines: `make setup-gotools test`

michelleN replied:
yep. thought about that too. will do :+1:

sgoings commented on expandybird/Makefile:
```diff
@@ -47,3 +49,21 @@ clean:
+
+.PHONY: vet
+vet:
+	$(GO_VET)
```
These commands actually could all live in include.mk for less duplication. :smile:

sgoings commented on include.mk:
```diff
@@ -0,0 +1,4 @@
+SETUP_GOTOOLS := @echo "Installing golint and vet" && go get -u github.com/golang/lint/golint && go get -u -v golang.org/x/tools/cmd/vet
+GO_VET := @echo "Running go vet ..." && go vet ./...
+GO_LINT := @echo "Running golint ..." && golint ./...
+GO_TEST := @echo "Running tests ..." && go test -v ./...
```
If you move the make targets [test, test-unit, lint, vet, setup-gotools] into this file, you'll probably no longer have a need to create make targets inside of make variables.

Commit:
makefile consolidations

Commit:
implement go vet suggestions

sgoings commented:
This looks lots better. Any reason why you didn't put the "test" target in the include.mk file so that it could just be inherited by all projects (sub or root)?

@jackgr I think this boosts deployment-manager contributor and maintainer UX - if someone just works on expandybird they're able to test, vet, lint, etc. from within the expandybird directory just like they would at the root or in another subproject. @michelleN also removed potential duplication (which is what you stated you were worried about), created a model for controlling build complexity down the road, and provided a consistent interface (namely: the test target which runs more than just unit tests now) for external contributors to rely on when they go to work on deployment-manager and internal developers to use even if we decide to tweak how certain subprojects run tests (or potentially get more advanced functional tests). What do you think?

sgoings commented on manager/Makefile:
```diff
@@ -12,6 +12,8 @@
 
+include ../include.mk
+
 .PHONY : all build test push container clean .project
```
This "test" PHONY target is no longer necessary as it's defined near the target below.

sgoings commented on resourcifier/Makefile:
```diff
@@ -14,6 +14,8 @@
 
+include ../include.mk
+
 .PHONY : all build test push container clean
```
See related comment in manager/Makefile re: duplicate `PHONY: test` statements

sgoings commented on .travis.yml:
```diff
@@ -9,4 +9,4 @@ install:
 
 script:
-  - make test
+  - make setup-gotools test
```
:+1:

sgoings commented:
Ooh ooh! Pretty please move the "info" target into the `include.mk` too?

jackgr commented:
Travis doesn't like it...

`The command "make setup-gotools test" exited with 2.`

jackgr commented:
/cc @michelleN @sgoings

michelleN commented:
@jackgr Fixed some issues raised by go vet. We should be in business now. What do you think?

jackgr commented:
Nice work, @michelleN!

jackgr commented:
Fixes #211.
